keras_cv/models/object_detection/retina_net/retina_net.py

```diff
@@ -415,27 +423,50 @@ def train_step(self, data):
         gradients = tape.gradient(total_loss, trainable_vars)
         self.optimizer.apply_gradients(zip(gradients, trainable_vars))
 
-        return self.compute_metrics(x, {}, {}, sample_weight={})
+        if not self._has_user_metrics:
+            return super().compute_metrics(x, {}, {}, sample_weight={})
+
+        y_pred = self.decode_predictions((box_pred, cls_pred), x)
+        return self.compute_metrics(x, y, y_pred, sample_weight=None)
 
     def test_step(self, data):
         x, y = unpack_input(data)
-
-        y = bounding_box.convert_format(
+        y_for_label_encoder = bounding_box.convert_format(
             y,
             source=self.bounding_box_format,
             target=self.label_encoder.bounding_box_format,
             images=x,
         )
-        boxes, classes = self.label_encoder(x, y)
+        boxes, classes = self.label_encoder(x, y_for_label_encoder)
         boxes = bounding_box.convert_format(
             boxes,
             source=self.label_encoder.bounding_box_format,
             target=self.bounding_box_format,
             images=x,
         )
-        _ = self.compute_loss(x, boxes, classes, training=False)
 
-        return self.compute_metrics(x, {}, {}, sample_weight={})
+        box_pred, cls_pred = self(x, training=False)
+        _ = self.compute_loss(x, box_pred, cls_pred, boxes, classes)
+
+        if not self._has_user_metrics:
+            return super().compute_metrics(x, {}, {}, sample_weight={})
+        y_pred = self.decode_predictions((box_pred, cls_pred), x)
+        return self.compute_metrics(x, y, y_pred, sample_weight=None)
+
+    def compute_metrics(self, x, y, y_pred, sample_weight):
```

Comment:
Why override this?

We should be able to just pass `metrics` to `super().compile` and the default metrics implementation should handle it.

See e.g. https://colab.sandbox.google.com/drive/1lKXIucTDaQ_50ih6d1S_kYdz31lnKHTG?resourcekey=0-3n7ckwRR1393vs0T3WP2Bw#scrollTo=j-m3LsD42lOU where I added metric support to RetinaNet without this change.

Reply:
you were only able to get this to work because we used the old box format at the time:

```
        return self.compute_metrics(x, tf.concat([y["boxes"], y["classes"]], axis=-1), preds, sample_weight={})
```

When you use bounding boxes the default `compute_metrics()` implementation interprets dictionaries as a mapping of distinct outputs.

Reply:
Unfortunately with dictionary boxes this is required.   Maybe we can come up with a utility or abstraction to make doing this easier?